### PR TITLE
Feature char v2

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by sopel39** [comment]
Supersedes: #5358 

Rebased on newest master.
Changes: Re2J and Joni casts are always registered. Only functions are selected with config property.
This is because for all registered casts (VARCHAR->Joni|Re2J, CHAR->Joni|Re2J) there should be casts registered.

FYI: @cberner

**Comment by cberner** [comment]
Note to self: review https://github.com/prestodb/presto/pull/5377 first, so we can base this work on top

**Comment by cberner** [comment]
Can you rebase this on @fiedukow 's work? I just accepted that, so it should be merged soon

**Comment by sopel39** [comment]
I have rebased on top of current master. I have also included https://github.com/prestodb/presto/pull/5583 commit and used in char code as requested by @cberner.

@cberner could you review the code. I think it is more important than EXPLAIN ANALYZE review because it touches wide codebase. Rebasing it every sprint became quite cumbersome. Thanks!

**Comment by cberner** [comment]
How about the aggregation spilling PR? Can you talk to @pnowojski about which is more important?

**Comment by sopel39** [comment]
We decided with @mattsfuller that we should go with char first. I think it should be fairly easy to review since it doesn't involve major design questions. However, maintaining it takes more and more time.

**Comment by cberner** [comment]
Cool

**Comment by fiedukow** [comment]
Seems like the constructors like `CHAR 'aaa'` dont work as expected.
Also I think we may make a use from cast `CHAR` to `VARBINARY` operator.

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+
+    private CharType(long length)
+    {
+        super(
+                new TypeSignature(
+                        StandardTypes.CHAR,
+                        singletonList(TypeSignatureParameter.of(length))),
+                Slice.class);
+
+        if (length < 1 || length > MAX_LENGTH) {
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: Why is the minimum `1` rather than `0`? `varchar(x)` allows `varchar(0)`

File: presto-main/src/main/java/com/facebook/presto/type/CharOperators.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+    private CharOperators() {}
+
+    private static SqlScalarFunction charLogicalOperator(OperatorType operatorType, String... methods)
+    {
+        Signature signature = Signature.builder()
+                .kind(SCALAR)
+                .operatorType(operatorType)
+                .argumentTypes(
+                        parseTypeSignature("char(x)", ImmutableSet.of("x")),
+                        parseTypeSignature("char(y)", ImmutableSet.of("y")))
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/CharOperators.java]: This should be `char(x)` and `char(x)`. All the comparison operators need to have the form `op(T,T)` or `between(T,T,T)` in the case of `BETWEEN`. See discussion around https://github.com/prestodb/presto/issues/5671

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java [hunk 3]
```diff
@@ -269,4 +271,16 @@ public FeaturesConfig setRegexLibrary(RegexLibrary regexLibrary)
         this.regexLibrary = regexLibrary;
         return this;
     }
+
+    public boolean isCharPadSpaces()
+    {
+        return charPadSpaces;
+    }
+
+    @Config("char.pad-spaces")
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Can you add a description of what this config does? It looks like `true` means implicitly pad all `char(x)` with spaces on the right, whereas `false` effectively treats them as `varchar(x)`

Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Also, do we need this to be configurable?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+
+        // preallocate the result
+        int bufferSize = slice.length() + resultLength - textLength;
+        Slice buffer = Slices.allocate(bufferSize);
+
+        // fill in the existing string
+        buffer.setBytes(0, slice);
+
+        // fill padding spaces
+        for (int i = slice.length(); i < bufferSize; ++i) {
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java]: Shouldn't this be dependent on the `char.pad-spaces` property?

File: presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+package com.facebook.presto.type;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.PARAMETER})
+public @interface FromLiteralParameter
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java]: Lets name this `LiteralParameter` to match the naming of `@TypeParameter`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java [hunk 6]
```diff
@@ -253,9 +254,26 @@ private TypeImplementationDependency(String signature)
...
         {
-            return typeManager.getType(SignatureBinder.bindVariables(signature, new BoundVariables(types, ImmutableMap.of())));
+            return typeManager.getType(SignatureBinder.bindVariables(signature, boundVariables));
+        }
+    }
+
+    private static final class LiteralImplementationDependency
+            implements ImplementationDependency
+    {
+        private final String literal;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java]: Lets name this `variable` or `literalName` it sounds like it's the value of the literal the way it is now

**Comment by cberner** [comment]
* "Add FromLiteralParamter as ScalarImplementation dependency": typo in commit message, and some minor comments. Otherwise it looks good

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpCasts.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+
+public final class JoniRegexpCasts
+{
+    private JoniRegexpCasts()
+    {
+    }
+
+    @ScalarOperator(OperatorType.CAST)
+    @SqlType(JoniRegexpType.NAME)
+    public static Regex joniRegexp(@SqlType(StandardTypes.VARCHAR) Slice pattern)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpCasts.java]: This should take `varchar(x)`

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java [hunk 8]
```diff
@@ -204,7 +205,7 @@ public void smokedTest()
         assertExecute("bound_pattern", VARCHAR, "%el%");
         assertExecute("bound_null_string", VARCHAR, null);
         assertExecute("bound_timestamp_with_timezone", TIMESTAMP_WITH_TIME_ZONE, new SqlTimestampWithTimeZone(new DateTime(1970, 1, 1, 0, 1, 0, 999, DateTimeZone.UTC).getMillis(), TimeZoneKey.getTimeZoneKey("Z")));
-        assertExecute("bound_binary_literal", VARBINARY, new SqlVarbinary(new byte[]{(byte) 0xab}));
+        assertExecute("bound_binary_literal", VARBINARY, new SqlVarbinary(new byte[] {(byte) 0xab}));
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java]: Whitespaces changes should be in a separate commit

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java [hunk 9]
```diff
@@ -312,6 +314,8 @@ private static Object fixValue(TypeSignature signature, Object value)
             case INTERVAL_DAY_TO_SECOND:
             case DECIMAL:
                 return String.class.cast(value);
+            case CHAR:
+                return Strings.padEnd(String.class.cast(value), getCharTypeLength(signature), ' ');
```
Comment (cberner) [inline presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: Padding on the client side doesn't seem like the right thing to do to me, especially since padding is a configurable option in the server

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 10]
```diff
@@ -353,6 +369,8 @@ public void addParametricType(ParametricType parametricType)
                         return Optional.of(getType(new TypeSignature(resultTypeBase)));
                     case StandardTypes.VARCHAR:
                         return Optional.of(createVarcharType(0));
+                    case StandardTypes.CHAR:
+                        return Optional.of(createCharType(1));
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Why `1` not `0`?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+import java.util.Objects;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.String.format;
+import static java.util.Collections.singletonList;
+
+public final class CharType
+        extends AbstractVariableWidthType
+{
+    public static final int MAX_LENGTH = 65_536;
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: Why limit to 64k?

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    {
+        if (block.isNull(position)) {
+            return null;
+        }
+
+        return block.getSlice(position, 0, block.getLength(position)).toStringUtf8();
+    }
+
+    @Override
+    public boolean equalTo(Block leftBlock, int leftPosition, Block rightBlock, int rightPosition)
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: The commit which defines the operators should be squashed into this one. It's required that comparable types have `EQUAL` and `HASH_CODE` and orderable must have the comparison operators

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    {
+        int leftLength = leftBlock.getLength(leftPosition);
+        int rightLength = rightBlock.getLength(rightPosition);
+        if (leftLength != rightLength) {
+            return false;
+        }
+        return leftBlock.equals(leftPosition, 0, rightBlock, rightPosition, 0, leftLength);
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: This has to be consistent with `EQUAL`. However, `EQUAL` changes its behavior based on padding option. We may want to just remove that config flag

**Comment by cberner** [comment]
* "Remove unused toSqlDecimal function": looks good
* "Make Joni and RE2J Regex cast always registered.": minor comments, otherwise looks good
* "Add test that ensures cast operators are available for coercible types": looks good

**Comment by cberner** [comment]
Have some basic questions about the behavior of `char`.

**Comment by cberner** [comment]
For reference, the Postgres docs on `char`: https://www.postgresql.org/docs/9.5/static/datatype-character.html

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 1 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: SQL standard specifies that minimum CHAR length is 1. See: https://mariadb.com/kb/en/sql-99/character-string-data-types/

File: presto-main/src/main/java/com/facebook/presto/type/CharOperators.java (hunk 2 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/CharOperators.java]: If I define all comparison operators as `op(T,T)`, then it will not be possible to compare `CHARs` with `NO PADDING` (default) option, e.g: `CHAR(5) 'a    '` would be equal to `CHAR(1) 'a'`. This is inconsistent with how `VARCHARs` are compared.

 BTW: since https://github.com/prestodb/presto/pull/5717 is merged I think `op(char(x), char(y))` should work now.

Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/CharOperators.java]: I have also re-factored this code to use FromLiteralParameter annotation

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Added description.

`PAD SPACES` is part of collation. We can assume just one collation without `PAD SPACES`, however it is useful to have this option selectable especially when Presto is being queried from other database (e.g. Teradata). Different databases do strings comparison differently, therefore with this option we can adjust for their behavior.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java (hunk 4 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/CharacterStringCasts.java]: `PAD SPACES` only affects how char strings are compared not how they are represented.

File: presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java]: I am not sure, I think it will be confused with `@LiteralParameters` annotation, which just specifies which parameters are literal. What do you think @fiedukow

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java (hunk 6 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java]: renamed to `literalName`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpCasts.java (hunk 7 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/JoniRegexpCasts.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java (hunk 8 above)
Comment (sopel39) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/TestExpressionCompiler.java]: this was fixed in some other patch, which is in master now

File: presto-client/src/main/java/com/facebook/presto/client/QueryResults.java (hunk 9 above)
Comment (sopel39) [inline presto-client/src/main/java/com/facebook/presto/client/QueryResults.java]: according to SQL standard, `CHAR(x)` always has `x` characters. Internally `CHAR(x)` is represented as variable length slice. However, when presenting to client it has to be padded with spaces to have length of `x`. We don't want to transmit fixed length `CHARs`, because spaces would waste bandwidth.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 10 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: char(0) is invalid type

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 11 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: When somebody creates table with `column CHAR(2000000)` and then casts `column` to `VARCHAR`, Presto would would append as most `2000000` spaces. Also on client side each row value would be appended with at most `2000000` spaces.

In general, `CHAR` type is usually used for short strings, e.g: labels, therefore there is no need for unlimited `CHAR`

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 12 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: squashed

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 13 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: equal here is always called on `CHARs` with the same length. The `PAD SPACES` flag only affects `CHARs` that have different length, e.g: `CHAR(5)` and `CHAR(10)`.

Logically, `CHAR(x)` has always `x` characters (with trailing spaces)

**Comment by sopel39** [comment]
Superseded by: https://github.com/prestodb/presto/pull/5501

File: presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java (hunk 5 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java]: There's already `TypeParameter` and `TypeParameters`, so I think we should keep things consistent

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 3 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: I'm not very familiar with SQL collation rules, but looking through the spec they look very complicated. Can you open an issue, so that we can discuss how collation should work in Presto, and start things off with your understanding of the spec?

File: presto-main/src/main/java/com/facebook/presto/type/CharOperators.java (hunk 2 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/CharOperators.java]: Lets discuss on https://github.com/prestodb/presto/issues/5800. Unless we decide that that's the wrong way forward, this isn't going to work

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 10 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Why invalid? The spec says "The value of a <length> shall be greater than 0 (zero).", but if we're following the spec then `varchar(0)` should be invalid too. I think it would be better to make it consistent and implement `char(0)` unless that causes an obvious problem

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 3 above)
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: Oh, I see there's already https://github.com/prestodb/presto/issues/5320 open

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 13 above)
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: Unfortunately, I don't think that's true. This method is used in JOINs and currently there's a type only coercions registered between `char` types, so I think you'll get the wrong answer if you JOIN with no padding

File: presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java (hunk 5 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/FromLiteralParameter.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java (hunk 3 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java]: I will move commit introducing `PAD SPACES` flag to the end, so that previous commits can be merged quicker

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 10 above)
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: ok, I will make it 0 min

File: presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java (hunk 13 above)
Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: So you are implying that if `type a` can be type only coerced to `type b` then in join operator `equalsTo` from either `type a` or `type b` would be used to compare block values (even though `type a != type b`)? I think that shouldn't be the case. If `type a` is type only coercible to `type b` that doesn't mean that `equalsTo` for those two types behaves identically (e.g coercion matters).

Consequently, this breaks VARCHAR-like comparison semantics for `CHAR` in JOINs. JOIN on `col1 CHAR(x=3)` and `col2 CHAR(y=5)` currently behaves like it first casts `col1` to `CHAR(5)`. For example:
```
presto> select * from (VALUES (CAST('a' AS CHAR(1)))) as t(a) JOIN (VALUES (CAST('a' AS CHAR(2)))) as t2(b) ON a = b;;
 a | b  
---+----
 a | a  
(1 row)
```
while:
```
presto> select * from (VALUES (CAST('a' AS CHAR(1)))) as t(a) JOIN (VALUES (CAST('a' AS CHAR(2)))) as t2(b) ON a = b WHERE a = b;
 a | b 
---+---
(0 rows)
```

I think `equalTo`, `hash` and `compareTo` that operate on `Blocks` from different types should be moved from Types in SPI and registered in `FunctionRegistry`. What do you think?

FYI: @losipiuk @fiedukow

Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: I agree that it shouldn't work like this, but unfortunately it does. @haozhun and I have been trying to clean this up in https://github.com/prestodb/presto/pull/5781 and https://github.com/prestodb/presto/pull/5778, but after that goes in more work needs to be done to remove "type only" coercions.

As I understand it, type only coercions have the following contract:
1) Implementation of the CAST must be the identity function.
2) All code in Presto, after analysis, must assume that type only coercions may be implicitly present.

I see a few ways forward here:
1) Add a flag like `experimental.enable-char-type` which would default to `false`, document all the known issues with `char(x)` and merge it, with that flag so that people can try it out.
2) Move all the `char(x)` features that don't work currently into a separate PR (coercions, `char(x)`/`char(y)` comparison...) and just merge the basics.
3) Fix the issues in our type system to support `char(x)` and then merge this PR afterwards.

Let me know which of those would work best for you.

Moving `equalTo`, `hash`, and `compareTo` on `Block` into the `FunctionRegistry` is something we've been talking about for a while. That's probably the best option, but one thing we should test first is whether those functions are actually necessary, or if the `EQUAL` operator could just be used by invoking the `MethodHandle` in a loop.

Comment (sopel39) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: > 2) All code in Presto, after analysis, must assume that type only coercions may be implicitly present.

Do you mean that type only coercion makes two types interchangeable? While their representation is the same, it doesn't necessarily mean that their semantics is the same too.

> I see a few ways forward here:

We will try some combination of 1) and 2) since it will let us merge the code faster. Maybe the best would be to add `experimental.enable-char-operators` flag so that the `CHAR` data can be just pipelined through Presto.
We are also happy to fix type system issues.

> but one thing we should test first is whether those functions are actually necessary

In case of primitive types those functions are probably not required. However, there will be performance impact in `Slice` based types, since `Slice` instances would have to be instantiated for each `EQUAL` call. That is why I think `Block` based function that are registered in `TypeManager` would be the best approach.

Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/type/CharType.java]: Lets move the discussion to https://github.com/prestodb/presto/issues/5320
